src/nccl_ofi_rdma.c

```diff
@@ -5503,9 +5468,24 @@ static int connect(nccl_net_ofi_ep_t *base_ep,
 
 		comm_state->stage = COMM_CONN_RESP_REQ_PENDING;
 
+	case COMM_CONN_RESP_REQ_PENDING:
+
+		// TODO: We might want to use READ_ONCE to read variable `connected'
+		if (!s_comm->connected) {
+			ret = ofi_process_cq(ep);
+
+			/*
+			 * To allow `connect` side to continue to process its cq,
+			 * don't return a valid send communicator to NCCL until
+			 * the connection is finalized.
+			 */
+			return ret;
```

Comment:
why `return`  indiscriminately? If the connect_response has been received, we want to return the scomm now, there is no need to wait another call to connect. right?

Reply:
That's right. We can check `s_comm->connected` again here. But even if we return indiscriminately, we will simply see that the connection is finalized on the next call to `connect`.

Reply:
exactly, so I say it is better to not postpone it

Reply:
Rearranged the logic for this to make more sense.